### PR TITLE
PEP 458: Add Sections for Signing Metadata and the Formats of Key Objects and Metadata

### PR DESCRIPTION
**Pull request by vladimir-v-diaz**
* Discuss the actual format of the JSON metadata files, and reference the example metadata files included in the "pep-0458-repository" subdirectory available here.

* Explain how the JSON metadata is signed and the signing functions provided by the repository management tool.  Code snippets showing the signing and writing of metadata are provided as well.

* Finally, a section for the password-protected, encrypted key files is added.  In addition to the cryptographic primitives that may be used, the format of the key objects stored in these encrypted key files and metadata is outlined.

### CODE REVIEW

File: pep-0458-tuf-online-keys.rst [hunk 1]
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
 __ https://github.com/theupdateframework/tuf/tree/develop/tuf#delegations
 
 
+File Formats of the PyPI JSON Metadata
+--------------------------------------
+
+This section presents the format of the JSON metadata files.  Examples of the
+roles and their formats mentioned here are available for review in the
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: "and their formats mentioned here" seems a bit off. Can you just delete "mentioned here"?

File: pep-0458-tuf-online-keys.rst [hunk 2]
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
...
+
+
+root.JSON
+~~~~~~~~~
+
+The root.json file is signed by the root role's keys.  It indicates
+which keys are authorized for all top-level roles, including the root
+role itself.  Revocation and replacement of top-level role keys, including
+for the root role, is done by changing the keys listed for the roles in
+this file.
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: This sentence seems pretty convoluted.

File: pep-0458-tuf-online-keys.rst [hunk 3]
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
...
+  }
+}
+
+VERSION is an integer that is greater than 0.  Clients MUST NOT replace a
+metadata file with a version number less than the one currently trusted.
+
+EXPIRES determines when metadata should be considered expired and no longer
+trusted by clients.  Clients MUST NOT trust an expired file.
+
+A ROLE is one of "root", "snapshot", "targets", "timestamp", or "mirrors".
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: 438: rather than "is one of" I suggest "may be"

File: pep-0458-tuf-online-keys.rst [hunk 4]
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
...
+A role for each of "root", "snapshot", "timestamp", and "targets" MUST be
+specified in the key list. The role of "mirror" is optional.  If not
+specified, the mirror list will not need to be signed if mirror lists are
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: line 441: "the mirror list will not need to be signed" add "even" at the end of this phrase

File: pep-0458-tuf-online-keys.rst [hunk 5]
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The targets.json metadata file lists the hashes and sizes of target files.
+Target files are the actual files that clients are intending to download (for
+example, the software updates they are trying to obtain).
+
+This file can optionally define other roles to which it delegates trust.
+Delegating trust means that the delegated role is trusted for some or all of
+the target files available from the repository. When delegated roles are
+specified, they are specified in a similar way to how the Root role specifies
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: I would change this "they are specified in a similar way to how" to this: "it is done similar to how"

File: pep-0458-tuf-online-keys.rst [hunk 6]
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
...
+  ...
+  ]
+}
+
+In order to discuss target paths, a role MUST specify only one of the
+"path_hash_prefixes" or "paths" attributes, each of which we discuss next.
+
+The "path_hash_prefixes" list is used to succinctly describe a set of target
+paths. Specifically, each HEX_DIGEST in "path_hash_prefixes" describes a set of
+target paths; therefore, "path_hash_prefixes" is the union over each prefix of
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: "union over" seems odd. should it be "union of"?

File: pep-0458-tuf-online-keys.rst [hunk 7]
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
...
+
+{
+  "_type" : "Timestamp",
+  "version" : VERSION,
+  "expires" : EXPIRES,
+  "meta" : METAFILES
+}
+
+METAFILES is the same is described for the snapshot.json file.  In the case of
+the timestamp.json file, this will commonly only include a description of the
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: 596: I would move "only" to after "include"

File: pep-0458-tuf-online-keys.rst [hunk 8]
```diff
@@ -395,6 +395,208 @@ specifying `delegations`__.
...
+The format of the timestamp file is as follows:
+
+{
+  "_type" : "Timestamp",
+  "version" : VERSION,
+  "expires" : EXPIRES,
+  "meta" : METAFILES
+}
+
+METAFILES is the same is described for the snapshot.json file.  In the case of
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: 595: "metafiles is the same" ____ (what?) then, use "is" or "as" depending on what works

File: pep-0458-tuf-online-keys.rst [hunk 9]
```diff
@@ -499,16 +701,143 @@ SHOULD also be made available to clients.
...
+Cryptographic keys MAY be stored in password-protected, encrypted key files.
+Administrators MAY use the repository tool to encrypt key files with
+AES-256-CTR-Mode and strengthen passwords with PBKDF2-HMAC-SHA256 (100K
+iterations by default, but this may be overridden by the developer). The
+current Python implementation of TUF can use any cryptographic library
+(PyCrypto [24]_ is currently used to encrypt the TUF key files, but support for
+PyCA Cryptography can be added in the future), may override the default number
+of PBKDF2 iterations, and the KDF tweaked to preference.  However, the exact
+cryptographic constructions can be adjusted to include future primitives added
+by the cryptographic libraries supported.
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: 735: something is wrong with "supported" - at the very least it is in the wrong place.

File: pep-0458-tuf-online-keys.rst [hunk 10]
```diff
@@ -499,16 +701,143 @@ SHOULD also be made available to clients.
...
+
+
+Keys objects stored in encrypted key files and in metadata have the format:
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: 738: should the first "Keys" be Key?

File: pep-0458-tuf-online-keys.rst [hunk 11]
```diff
@@ -499,16 +701,143 @@ SHOULD also be made available to clients.
...
+The repository management tool may be used to generate the cryptographic keys
+and sign the PyPI metadata downloaded by end users.  The following Python code
+demonstrates how to generate and import cryptographic keys with the repository
+management tool:
+
+.. code-block::
+
+  >>> from tuf.repository_tool import *
+
+  # Generate and write an ed25519 key pair.  The private key is saved encrypted.
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: 786: "saved encrypted" doesn't seem right. Maybe "encrypted before it is saved"

File: pep-0458-tuf-online-keys.rst [hunk 12]
```diff
@@ -499,16 +701,143 @@ SHOULD also be made available to clients.
...
+
+  # Import the ed25519 public key just created . . .
+  >>> public_ed25519_key = import_ed25519_publickey_from_file('/path/to/ed25519_key.pub')
+
+  # and its corresponding private key.
+  >>> private_ed25519_key = import_ed25519_privatekey_from_file('/path/to/ed25519_key')
+  Enter a password for the encrypted ED25519 key: 
+
+
+The repository tool can use the cryptographic keys that are imported to sign
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: 800: do you mean to say "use the cryptographic keys that are imported to sign" or, could you instead say (better): "use the imported cryptographic keys to sign"

File: pep-0458-tuf-online-keys.rst [hunk 13]
```diff
@@ -499,16 +701,143 @@ SHOULD also be made available to clients.
...
+    { "keyid" : KEYID,
+      "method" : METHOD,
+      "sig" : SIGNATURE,
+    },
+    ...
+  ]
+}
+
+ROLE is a dictionary whose "_type" field describes the role type.
+KEYID is the identifier (64-byte hexstring) of the key signing the ROLE dictionary.
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: change "of the key signing" to "of the key that signs"

File: pep-0458-tuf-online-keys.rst [hunk 14]
```diff
@@ -571,19 +900,19 @@ Project developers expect the distributions they upload to PyPI to be
...
 many readers and writers simultaneously access the same metadata and
 distributions.  That is, there needs to be a way to ensure consistency of
-metadata and repository files when multiple developers simulaneously change the
-same metadata or distributions.  There are also issues with consistency on PyPI
-without TUF, but the problem is more severe with signed metadata that MUST keep
-track of the files available on PyPI in real-time.
+metadata and repository files when multiple developers simultaneously update
+the same metadata or distributions.  Without TUF, there are also issues with
+consistency on PyPI, but the problem is more severe with signed metadata that
+MUST keep track in real-time the files available on PyPI.
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: 906: revise "keep track in real-time the files" to "keep track, in real-time, of the files"

File: pep-0458-tuf-online-keys.rst [hunk 15]
```diff
@@ -571,19 +900,19 @@ Project developers expect the distributions they upload to PyPI to be
...
-arbitrary metadata injected by an attacker.  The problem would also occur for
-mirrors attempting to sync with PyPI.
+Suppose that PyPI generates a *snapshot*, which describes the latest version of
+every metadata (except *timestamp*), at specified version 1, and that a client
+requests this *snapshot* from PyPI.  While the client is busy downloading this
+*snapshot*, PyPI timestamps a new snapshot at, say, version 2.  Without
+ensuring consistency of metadata, the client would find itself with a copy of
+*snapshot* that is inconsistent with what is available on PyPI: this situation
+is indistinguishable from arbitrary metadata injected by an attacker.  The
+problem would also occur with mirrors attempting to sync with PyPI.
```
Comment (lvigdor) [inline pep-0458-tuf-online-keys.rst]: 915: change "attempting to" to "that attempt to"

**Comment by vladimir-v-diaz** [comment]
@lvigdor 
Thanks you for reviewing this pull request.  Addressed all of your comments here: https://github.com/vladimir-v-diaz/interoperability-peps/commit/c838e229aed78df274e8a867b2877609247e172e

**Pull request merged**